Calculate and print stack size on AVR builds.

purdeaandrei opened the pull request:
## Description

AVRs have very limited RAM available. It is nice to have at the end of the build a message showing how much flash space is left, but what is even more important is knowing how we are with regard to RAM usage. Of course linking will fail if so much ram is used up by global variables, that it won't fit even with a 0-sized stack, however what is not detected and cannot be detected at build time is the case when the global variables fit in RAM, but the stack starts overriding them, because not enough stack space has been left. The best thing we can do for this is to calculate and print the stack size, and have the user interpret this.

## Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply. -->
- [x] Core
- [ ] Bugfix
- [x] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

sigprof commented:
> Of course linking will fail if so much ram is used up by global variables, that it won't fit even with a 0-sized stack

Actually it won't fail even in that case, unless the ancient gcc-5.4-based toolchain with Atmel patches (currently found in Debian and Ubuntu) is used.  Apparently all other toolchains just set the .data section size to the maximum supported by the architecture (slightly less than 64K), therefore the ld check for overflow does not work.

purdeaandrei commented:
> it won't fail even in that case, unless the ancient gcc-5.4-based toolchain with Atmel patches (currently found in Debian and Ubuntu) is used

Interesting, that's exactly what I've been using. Just tested my current code with 12.1.0 too, and it fails in that case because it takes a negative number `& 0xffff`, I'll convert to a draft for now.

Btw gcc 12.1.0 consumes 8 bytes more than gcc 5.4.

This is probably not related, but other weirdness I've seen with fresh gcc in the past was that when compiling TMK it resulted in completely broken binaries, with much smaller file size. Didn't really have time to debug that one, maybe it's just a bug in TMK, in any case that's one of the reasons I've been avoiding fresh gcc.

sigprof commented:
> Just tested my current code with 12.1.0 too, and it fails in that case because it takes a negative number `& 0xffff`, I'll convert to a draft for now.

The same thing happens in NixOS (avr-gcc  8.4.0, `_end` is `00800380`, while `__stack` is `00000aff`); however, the actual shell calculation seems to work.

> This is probably not related, but other weirdness I've seen with fresh gcc in the past was that when compiling TMK it resulted in completely broken binaries, with much smaller file size.

Yes, there are some issues — e.g., Fedora 38, avr-gcc 12.2.0 apparently miscompiles some RGB Matrix code when LTO is enabled: https://discord.com/channels/440868230475677696/440868230475677698/1098993586550345790

purdeaandrei commented:
@sigprof

> The same thing happens in NixOS (avr-gcc 8.4.0, `_end` is `00800380`, while `__stack` is `00000aff`); however, the actual shell calculation seems to work.

That `008` prefix I think it's just used for categorizing linkable symbols, or something like that, that's why I added `& 0xffff` everywhere, that prefix shows up in old GCC as well. I was referring to the case when for example _end is `00800b20`, in that case the result will be `(0xaff + 1 - 0xb20) & 0xffff = 65504` In which case we should fail with an error.

Anyway, I have now fixed my code, and made it behave more like the firmware size check. Except I have NOT added a warning that triggers at a limit, cause it's not clear what the limit should be. I could measure how much stack an example firmware uses by paiting the stack in crt0, but that would not be applicable to other firmwares, for example VIA/VIAL, and custom matrix firmwares will all likely use different amounts of stack, so I wouldn't know what default to set such limit to.

So right now there's two conditions:
1) Fresh GCC can trigger an error like this, if RAM overflows (i.e. negative amount of stack space left):
![image](https://user-images.githubusercontent.com/544648/233796646-6613f981-18eb-4b93-9021-dd4ce9c6d16d.png)
2) Old GCC will print a linker error instead on RAM overflow:
![image](https://user-images.githubusercontent.com/544648/233796952-0ae8a0c4-a4f1-456f-a864-b3f8cc99d795.png)

3) In normal cases any GCC will print something like the following:
![image](https://user-images.githubusercontent.com/544648/233796687-42053b16-a920-4e39-bb85-039254dd487f.png)
And it's the users responsibility to know if the amount is enough.
Also note that in this case we're not printing an [OK] either, it's just an info message.

On a second thought maybe I should still set up a warning message that would be only triggered if the keyboard/keymap configures a limit in their rules.mk?

Please review.

purdeaandrei commented:
Not stale, please review!

tzarc commented:
Will need conflict resolution but for the most part looks okay.

purdeaandrei commented:
Rebased, and conflicts resolved

purdeaandrei commented:
I made a new PR cause I wasn't able to re-open this one: https://github.com/qmk/qmk_firmware/pull/23101
